### ISSUE TITLE
[FR] Taking REPL modes seriously

### ISSUE THREAD

**Issue by BeastyBlacksmith** [JuliaLang/julia#31910]
As discussed in https://discourse.julialang.org/t/allow-escape-key-to-leave-shell-or-pkg-mode/23802 and also relates to (https://github.com/JuliaLang/julia/issues/19202)
It would be good to be able to choose between persistent and single execution REPL mode in a consistent manner. 

My proposal is to have single mode character ( `?`, `;`, `]` ) be single execution mode and meta + mode character be persistent mode.

Furthermore the possibility to get back to julia REPL with cursor mid-line with e.g. `esc` would also take account of inattention of users.

**Comment by stevengj** [comment]
👎  on making single-execution the default, but 👍 on making esc go back to default mode.

**Comment by BeastyBlacksmith** [comment]
Could be the other way around, too. Or every mode but help mode is persistent by default, but you can toggle the other.
Wouldn't special case more than one mode, though.

**Comment by KristofferC** [comment]
Having extra keybinds for entering sticky vs non-stricky seems like over complicating things. Having the default be sticky and allowing `esc` to exit to default while keeping the current text in the prompt seems good enough.

>  Do we need a poll for this?

There is no need for polls.

**Comment by GunnarFarneback** [comment]
Note that `esc` does things already, although it's probably rarely really needed. Personally I wouldn't use `esc` to leave a mode anyway since the ergonomics of that key is worse than `ctrl+A` followed by `backspace`.

**Comment by JeffBezanson** [comment]
Agreed; if you don't like sticky repl modes (and I'm not a fan of them, but it's not a big deal) adding `esc` doesn't really change things much.

**Comment by stevengj** [comment]
@GunnarFarneback, it's ctrl-a followed by backspace followed by ctrl-e to get back to your previous editing position.  Not to mention that for users not raised on Unix terminals the ctrl-a and ctrl-e shortcuts are not particularly intuitive or discoverable, whereas using esc to "escape" is perhaps easier to guess.

**Comment by juliohm** [comment]
Could you please clarify what was agreed? Currently in Julia v1.2 the only mode that is persistent is package mode, correct? There are plans to make the other modes (help and shell modes) sticky as well? Maybe a feature for Julia v1.3?

**Comment by PetrKryslUCSD** [comment]
I have an objection to making the modes sticky and having a keystroke for jumping out of the stickiness.
The problem is the keystroke does not show up in the history. Therefore it is difficult to mix the modes.
https://discourse.julialang.org/t/what-do-people-think-about-the-package-mode-being-persistent/45527/7

**Comment by PetrKryslUCSD** [comment]
I think a better solution would be  to have a command for jumping is out of the stickiness. Even better not to haveany sticky modes at all .

**Comment by KristofferC** [comment]
> The problem is the keystroke does not show up in the history. Therefore it is difficult to mix the modes.

The REPL mode does show up in history:

```
# time: 2019-09-22 09:40:32 CEST
# mode: pkg
	add Revise
# time: 2019-09-22 09:40:52 CEST
# mode: julia
	using Revise
```

**Comment by PetrKryslUCSD** [comment]
True, and I do use this information. The problem with printing the history of commands is that there is no way I can think of to indicate how to jump out of the sticky package mode.

Let us say I do:
```
]activate .
instantiate
[backspace]
include("make.jl")
```
Printing
```
]activate .
]instantiate
include("make.jl")
```
will have missed the end of the sticky mode, and the `include` will fail.

**Comment by KristofferC** [comment]
What is your goal? Being able to paste a REPL session including swapping between terminal modes?

**Comment by PetrKryslUCSD** [comment]
That is right. I want to be able to recover the sequence of commands
to put into a script or a function. Or just copy-and-paste to execute again.

Petr Krysl
Prof. and Vice chair for undergraduate education
Structural engineering department
University of California, San Diego

**Comment by KristofferC** [comment]
Okey, already on mac and Linux you can now paste

```
julia> 1+1
2

julia> print("hello")
```

into the REPL and it will work. That could pretty easily be extendible for `shell>` and `... pkg>`.

**Comment by PetrKryslUCSD** [comment]
How does that help?

**Comment by KristofferC** [comment]
Okay, I read your comment a bit too fast. I thought you wanted something that could be copy-pasted.

> I want to be able to recover the sequence of commands to put into a script or a function. 

So that would be Julia code then? The REPL modes are special in that they don't accept Julia code so I don't really see how this would work.

**Comment by PetrKryslUCSD** [comment]
> So that would be Julia code then? The REPL modes are special in that they don't accept Julia code so I don't really see how this would work.

That is precisely my point. I don't know if we are on the same page, but the history handling already works fine for the shell mode and the help mode.  I simply prefix those modes either with ; or with ?. But the package mode is sticky, so one closing bracket will throw it into a mode where a keystroke is required. There is no indication that I can think of to indicate that in a printout of the history, it is not a command, it is an invisible keystroke.

**Comment by KristofferC** [comment]
> That is precisely my point. I don't know if we are on the same page, but the history handling already works fine for the shell mode and the help mode. I simply prefix those modes either with ; or with ?. 

That only works in Windows right now (since it doesn't support bracket pasting). For example, me pasting

```
;ls 
```

will just get me

```
julia> ;ls
```

Also, this means that you do want to be able to paste the output, not being able to put it into a function...? How would

```
function my_session()
    ;ls
end
```

work?

**Comment by PetrKryslUCSD** [comment]
>That only works in Windows right now (since it doesn't support bracket pasting). 

I see. My bad, I haven't actually tested on Linux yet. That is a bummer.
When you say "right now", does that imply that this will be gone at one point? Or rather that it will be enabled on the Mac and on Linux?

**Comment by KristofferC** [comment]
Well, it would work if https://github.com/JuliaLang/julia/issues/31910#issuecomment-680896108 would be extended since you could then paste:

```
julia> 1+1

shell> ls

pkg> st
```

**Comment by PetrKryslUCSD** [comment]
Is this supposed to work universally? I tried on Windows without luck.

**Comment by PetrKryslUCSD** [comment]
Would it also work like this?
```
pkg> activate .
pkg> st
```

**Comment by LilithHafner** [comment]
From triage, 
- Non-sticky by default: no
- esc to exit:
  - Hits difficult terminal issues
  - We already have backspace and control-c
